cranelift/codegen/src/alias_analysis.rs

```diff
@@ -0,0 +1,387 @@
+                        };
+                        log::trace!(
+                            "alias analysis: at inst{}: load with last_store inst{} at loc {:?}",
+                            inst.index(),
+                            last_store.map(|inst| inst.index()).unwrap_or(usize::MAX),
+                            mem_loc
+                        );
+
+                        // Is there a Value already known to be stored
+                        // at this specific memory location?  If so,
+                        // we can alias the load result to this
+                        // already-known Value.
+                        //
+                        // Check if the definition dominates this
+                        // location; it might not, if it comes from a
+                        // load (stores will always dominate though if
+                        // their `last_store` survives through
+                        // meet-points to this use-site).
+                        let aliased = if let Some((def_inst, value)) =
+                            self.mem_values.get(&mem_loc).cloned()
+                        {
+                            log::trace!(
+                                " -> sees known value v{} from inst{}",
+                                value.index(),
+                                def_inst.index()
+                            );
+                            if self.domtree.dominates(def_inst, inst, &pos.func.layout) {
+                                log::trace!(
+                                    " -> dominates; value equiv from v{} to v{} inserted",
+                                    load_result.index(),
+                                    value.index()
+                                );
+
+                                pos.func.dfg.detach_results(inst);
+                                pos.func.dfg.change_to_alias(load_result, value);
+                                pos.remove_inst_and_step_back();
+                                true
+                            } else {
+                                false
+                            }
+                        } else {
+                            false
+                        };
+
+                        // Otherwise, we can keep *this* load around
+                        // as a new equivalent value.
+                        if !aliased {
+                            log::trace!(
+                                " -> inserting load result v{} at loc {:?}",
+                                load_result.index(),
+                                mem_loc
+                            );
+                            self.mem_values.insert(mem_loc, (inst, load_result));
+                        }
+                    }
+                }
+
+                state.update(pos.func, inst);
+            }
+        }
+    }
+}
+
+fn get_ext_opcode(op: Opcode) -> Option<Opcode> {
+    match op {
```

Comment:
Just to double check ourselves

```suggestion
fn get_ext_opcode(op: Opcode) -> Option<Opcode> {
    debug_assert!(op.can_load() || op.can_store());
    match op {
```

Reply:
Done!